core/src/sumeragi/mod.rs

```diff
@@ -64,9 +65,25 @@ pub struct SumeragiHandle {
 }
 
 impl SumeragiHandle {
-    /// Pass closure inside and apply fn to [`WorldStateView`]
-    pub fn wsv<T>(&self, f: impl FnOnce(&WorldStateView) -> T) -> T {
-        f(&self.public_wsv.lock())
+    /// Pass closure inside and apply fn to [`WorldStateView`].
+    /// This function must be used with very cheap closures.
+    /// So that it costs no more than cloning wsv.
+    pub fn apply_wsv<T>(&self, f: impl FnOnce(&WorldStateView) -> T) -> T {
+        f(&self.public_wsv_receiver.borrow())
+    }
+
+    /// Get public clone of [`WorldStateView`].
+    /// Cloning of [`WorldStateView`] is cheap.
+    pub fn wsv(&self) -> WorldStateView {
```

Comment:
I would like this to be called `wsv_clone`. I believe you that cloning is cheap on the workloads we test for. But this clone scales O(n) with the size of the wsv and therefore with the user count of the blockchain. I'm not saying that we should prefer apply_wsv to wsv_clone or vice versa. Just that the name should reflect this difference.

Reply:
You're correct, cheep cloning is a leftover from persistent structures attempt. I will remove this comment and rename method.